### PR TITLE
Implement Data Decoder

### LINKED ISSUE #33: Add decoder
# What is needed?
Add decoder endpoints, for a given data and contract address return the data decoded.

### PR DESCRIPTION
**Pull request by Uxio0**
- Closes #33
Adapt decoder from `Safe Tx Service`:
  - Refactor methods and make it `async`
  - Use a cache compatible with async methods
  - Copy and refactor tests

### CODE REVIEW

File: app/datasources/db/models.py [hunk 1]
```diff
@@ -48,6 +53,17 @@ class Abi(SqlQueryBase, SQLModel, table=True):
     source: AbiSource | None = Relationship(back_populates="abis")
     contracts: list["Contract"] = Relationship(back_populates="abi")
 
+    @classmethod
+    async def get_abis_sorted_by_relevance(
+        cls, session: AsyncSession
+    ) -> AsyncIterator[ABI]:
+        """
+        :returns: Abi JSON, with the ones with less relevance first
```
Comment (moisses89) [inline app/datasources/db/models.py:61]:
```suggestion
        :return: Abi JSON, with the ones with less relevance first
```

File: app/datasources/db/models.py [hunk 2]
```diff
@@ -78,3 +94,17 @@ class Contract(SqlQueryBase, SQLModel, table=True):
...
+    async def get_abi_by_contract_address(
+        cls, session: AsyncSession, address: bytes
+    ) -> ABI | None:
+        results = await session.exec(
+            select(Abi.abi_json)
+            .join(cls)
+            .where(cls.address == address)
+            .where(cls.abi_id == Abi.abi_hash)
+        )
+        if result := results.first():
```
Comment (moisses89) [inline app/datasources/db/models.py:108]: In this case we will return an ABI for any chain, the ABI could be the not expected, could you add a comment with a TODO to add support for multichain?

File: app/services/data_decoder.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+    session: AsyncSession | None
+
+    async def init(self, session: AsyncSession) -> None:
+        """
+        Initialize the data decoder service, loading the ABIs from the database and storing the 4byte selectors
+        in memory
+        """
+        logger.info("%s: Loading contract ABIs for decoding", self.__class__.__name__)
+        self.fn_selectors_with_abis: dict[bytes, ABIFunction] = (
+            await self._generate_selectors_with_abis_from_abis(
```
Comment (moisses89) [inline app/services/data_decoder.py:78]: Would make sense add this as a field for the ABI? Not in this PR but for the future.

File: app/services/data_decoder.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+    @alru_cache(maxsize=2048)
+    async def get_contract_abi(self, address: Address) -> ABI | None:
+        """
+        Retrieves the ABI for the contract at the given address.
+
+        :param address: Contract address
+        :return: List of ABI data if found, `None` otherwise.
+        """
+        async with AsyncSession(get_engine(), expire_on_commit=False) as session:
```
Comment (moisses89) [inline app/services/data_decoder.py:138]: Why this function is not using getting the session from database functions? you can use the decorator too.

**Review by moisses89 — CHANGES_REQUESTED** [review]
Awesome work!
Would be nice to add at least a log when the signature and contract address is not matching perfectly, or add a TODO in that part of code because we will work in the not 100% match feature.

**Comment by Uxio0** [comment]
> Awesome work! Would be nice to add at least a log when the signature and contract address is not matching perfectly, or add a TODO in that part of code because we will work in the not 100% match feature.

I think insteaf of a `log` we should return information about that in the endpoint, like `fullMatch: True` or something like that

File: requirements/prod.txt [hunk 5]
```diff
@@ -1,10 +1,13 @@
 aio-pika==9.5.1
 alembic==1.14.0
+async-lru==2.0.4
 asyncpg==0.30.0
 dramatiq[redis, watch]==1.17.1
 fastapi[all]==0.115.6
 periodiq==0.13.0
 pydantic-settings==2.7.0
 redis[hiredis]==5.2.1
+safe-eth-py==6.0.0
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply requirements/prod.txt:10]: Sure, done

File: app/datasources/db/models.py (hunk 2 above)
Comment (Uxio0) [reply app/datasources/db/models.py:108]: Sure

**Comment by moisses89** [comment]
> > Awesome work! Would be nice to add at least a log when the signature and contract address is not matching perfectly, or add a TODO in that part of code because we will work in the not 100% match feature.
> 
> I think instead of a `log` we should return information about that in the endpoint, like `fullMatch: True` or something like that

Yeah, agree but I mean leave any comment in the code as a TODO to check in the future.

**Review by moisses89 — APPROVED** [review]

**Pull request merged**